### PR TITLE
WIP - Repartitioning improvement (Do not review yet)

### PR DESCRIPTION
**Pull request by yingsu00**
NOTE: This is a working in progress PR. Please do not review yet.

Before this change, the PartitionedOutputOperator write rows of each
partition to BlockBuilders first, then serialize them into SerializedPage
before appending them to OutputBuffer. This change will skip the
BlockBuilder building step and directly append values to the buffers.
We calculate the row sizes for each incoming page, and only write the rows
that would fit in the buffers. When the buffers are full, we flush them
to SerializedPage and send it out.

This commit introduces several new methods in Block interface because we
wanted to do batch processing in tight loops. There were existing study
showing gains working on raw arrays in "batch" mode in tight loop over
working on Blocks in "row" mode. We will do additional JMH benchmarks
to compare the effect for repartitioning case.

Please also note that this commit does not contain the buffer pool logic yet.
The buffers are with fixed size and are assumed to have enough space for
now.

You can use the following query to trace the code:
SELECT checksum(l.partkey) FROM lineitem l, partsupp p WHERE l.partkey = p.partkey AND l.suppkey = p.suppkey;

### CODE REVIEW

File: presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java [hunk 1]
```diff
@@ -295,4 +295,12 @@ default Block getLoadedBlock()
     {
         return this;
     }
+
+//    int[] getSizesInBytes();
+//
+//    int[] getSizesInBytes(int[] offsets, int offset, int length);
+
+    void accumulateSizesInBytes(int[] outputSizesInBytes);
+    void accumulateSizesInBytes(int[] upperLevelOffsets, int[] outputSizesInBytes);
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java]: I'd rename these methods to match `getRegionSizeInBytes` and `getPositionsSizeInBytes`.

```
void appendPositionSizesInBytes(int[] sizesInBytes);

void appendRegionSizesInBytes(int[] offsets, int[] sizesInBytes);
```

File: presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java [hunk 2]
```diff
@@ -295,4 +295,12 @@ default Block getLoadedBlock()
...
+    void appendValuesToBuffers(BlockEncodingBuffers blockEncodingBuffers);
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java]: I'd rename this method to `writeTo`

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java [hunk 3]
```diff
@@ -294,6 +299,70 @@ public Page getOutput()
...
+    {
+        private final int partition;
+        private final AtomicLong rowsAdded;
+        private final AtomicLong pagesAdded;
+        private final PagesSerde serde;
+        private final Lifespan lifespan;
+        private int[] positions;
+        private int[] rowSizes;
+        private int positionCount;
+        PageEncodingBuffers pageEncodingBuffers;
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java]: Initialize these variables here and remove initialization logic from the ctor:

```
        private final Lifespan lifespan;

        private int[] positions = new int[100];
        private int[] rowSizes = new int[100];
        private int positionCount;
        private PageEncodingBuffers pageEncodingBuffers = new PageEncodingBuffers();

        PartitionData(int partition, AtomicLong pagesAdded, AtomicLong rowsAdded, PagesSerde serde, Lifespan lifespan)
        {
            this.partition = partition;
            this.pagesAdded = pagesAdded;
            this.rowsAdded = rowsAdded;
            this.serde = serde;
            this.lifespan = requireNonNull(lifespan);
        }
```

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java [hunk 4]
```diff
@@ -294,6 +299,70 @@ public Page getOutput()
...
+        {
+            if (positionCount == 0) {
+                return;
+            }
+
+            if ((pageEncodingBuffers.isEmpty())) {
+                pageEncodingBuffers.prepareBuffers(page);
+            }
+
+            int rowsWritten = 0;
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java]: Use positions term consistently: rowsWritten  -> positionsWritten

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java [hunk 5]
```diff
@@ -399,6 +476,48 @@ public void partitionPage(Page page)
...
+            Arrays.fill(rowSizes, 0);
+            for (int i = 0; i < page.getChannelCount(); i++) {
+                page.getBlock(i).accumulateSizesInBytes(rowSizes);
+            }
+        }
+
+        private void populateRowNumbersForEachPartition(Page page)
+        {
+            for (PartitionData target : partitionData) {
+                target.positionCount = 0;
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java]: Add reset method to PartitionData:

```
        public void reset()
        {
            positionCount = 0;
        }

            for (PartitionData partition : partitionData) {
                partition.reset();
            }
```

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java [hunk 6]
```diff
@@ -399,6 +476,48 @@ public void partitionPage(Page page)
...
+                // TODO: This is still the row-based way. We need to change it to vectorized way if the profile shows gain.
+                int partition = partitionFunction.getPartition(partitionFunctionArgs, position);
+                PartitionData target = partitionData[partition];
+                if (target.positions.length <= target.positionCount) {
+                    target.positions = Arrays.copyOf(target.positions, 2 * target.positions.length);
+                    target.rowSizes = Arrays.copyOf(target.rowSizes, 2 * target.rowSizes.length);
+                }
+                target.positions[target.positionCount] = position;
+                target.rowSizes[target.positionCount] = rowSizes[position];
+                target.positionCount++;
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java]: Add appendPosition method to PartitionData:

```
        public void appendPosition(int position, int rowSize)
        {
            ensureCapacity();
            positions[positionCount] = position;
            rowSizes[positionCount] = rowSize;
            positionCount++;
        }

        private void ensureCapacity()
        {
            if (positions.length == positionCount) {
                positions = Arrays.copyOf(positions, 2 * positionCount);
                rowSizes = Arrays.copyOf(rowSizes, 2 * positionCount);
            }
        }

            for (int position = 0; position < page.getPositionCount(); position++) {
                // TODO: This is still the row-based way. We need to change it to vectorized way if the profile shows gain.
                int partition = partitionFunction.getPartition(partitionFunctionArgs, position);
                partitionData[partition].appendPosition(position, rowSizes[position]);
            }
```

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockEncodingBuffers.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+import static com.facebook.presto.spi.block.ByteArrayUtils.writeLengthPrefixedString;
+import static com.facebook.presto.spi.block.ColumnarArray.toColumnarArray;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static java.lang.Math.max;
+import static sun.misc.Unsafe.ARRAY_BYTE_BASE_OFFSET;
+
+public class ArrayBlockEncodingBuffers
+        extends BlockEncodingBuffers
+{
+    public static final String NAME = "LONG_ARRAY";
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockEncodingBuffers.java]: Use ArrayBlockEncoding.NAME

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffersFactory.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+public class BlockEncodingBuffersFactory
+{
+    private BlockEncodingBuffersFactory()
+    {}
+
+    static BlockEncodingBuffers createBlockEncodingBuffers(Block block)
+    {
+        switch (block.getEncodingName()) {
+            case "LONG_ARRAY":
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffersFactory.java]: Use LongArrayBlockEncoding.NAME

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffersFactory.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+                return new LongArrayBlockEncodingBuffers();
+            case "ARRAY":
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffersFactory.java]: Use ArrayBlockEncoding.NAME

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffersFactory.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+public class BlockEncodingBuffersFactory
+{
+    private BlockEncodingBuffersFactory()
+    {}
+
+    static BlockEncodingBuffers createBlockEncodingBuffers(Block block)
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffersFactory.java]: Would it make sense to move this method into BlockEncodingSerde ?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    public void copyPositions(int[] positions, int startIndex, int batchSize)
+    {
+        if (nextBatchOfPositions == null || nextBatchOfPositions.length < batchSize) {
+            nextBatchOfPositions = new int[batchSize * 2];
+        }
+        System.arraycopy(positions, startIndex, this.nextBatchOfPositions, 0, batchSize);
+        this.nextBatchSize = batchSize;
+    }
+
+    // beginPosition inclusive, endPosition exclusive
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java]: This comment won't be necessary if arguments are renamed to offset and length:

```
    public void appendPositionRange(int offset, int length)
    {
        if (nextBatchOfPositions == null || nextBatchOfPositions.length < length) {
            nextBatchOfPositions = new int[max(nextBatchOfPositions.length * 2, length)];
        }

        for (int i = 0; i < length; i++) {
            nextBatchOfPositions[nextBatchSize + i] = offset + i;
        }

        nextBatchSize = length;
    }
```

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    abstract void writeTo(SliceOutput sliceOutput);
+
+    abstract int getSizeInBytes();
+
+    public void updateBufferedRowCount(int newRowCount)
+    {
+        bufferedRowCount += newRowCount;
+    }
+
+    public void copyPositions(int[] positions, int startIndex, int batchSize)
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java]: Based on the usage of this method, I suggest to rename arguments to offset and length and add a Block argument:

```
    public void copyPositions(int[] positions, int offset, int length, Block block)
    {
        if (nextBatchOfPositions == null || nextBatchOfPositions.length < length) {
            nextBatchOfPositions = new int[length * 2];
        }
        System.arraycopy(positions, offset, this.nextBatchOfPositions, 0, length);
        this.nextBatchSize = length;
        block.appendValuesToBuffers(this);
        bufferedRowCount += length;
    }


    public void appendRows(Page page, int[] positions, int startIndex, int length)
    {
        for (int i = 0; i < page.getChannelCount(); i++) {
            blockEncodingBuffers[i].copyPositions(positions, startIndex, length, page.getBlock(i));
        }
        bufferedRowCount += length;
    }
```

File: presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+{
+    private int channelCount;
+    int bufferedRowCount;
+    BlockEncodingBuffers[] blockEncodingBuffers;
+
+    private int capacity;
+
+    public PageEncodingBuffers()
+    {
+        channelCount = 0;
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java]: These initializations are not needed:

```
    private int channelCount;
    private int bufferedRowCount;
    private BlockEncodingBuffers[] blockEncodingBuffers;
    private final int capacity = DEFAULT_MAX_PAGE_SIZE_IN_BYTES;

    public PageEncodingBuffers()
    {
    }
```

File: presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    public boolean isFull()
+    {
+        return capacity <= getSizeInBytes();
+    }
+
+    public void prepareBuffers(Page page)
+    {
+        capacity = DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java]: remove this

File: presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    public BlockEncodingBuffers getBlockEncodingBuffers(int position)
+    {
+        // TODO: validate position
+        return blockEncodingBuffers[position];
+    }
+
+    public int calculateNumberRowsFitInBuffers(int[] rowSizes, int numberOfRows, int firstRowIndex)
+    {
+        // TODO: validate rows and rowSizes length >= numberOfRows > firstRowIndex
+        int roomInBytes = capacity - getSizeInBytes();
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java]: Can this logic be simplified?

```
        int remainingBytes = capacity - getSizeInBytes();

        for (int i = startPosition; i < positionCount; i++) {
            remainingBytes -= positionSizesInBytes[i];
            if (remainingBytes <= 0) {
                return max(i - startPosition, 1);
            }
        }

        return positionCount;
```

File: presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+    // Copy the buffers in order into the final slice
+    public Slice serialize()
+    {
+        // TODO: Validate the bufferedRowCount > 0
+        //  The performace using DynamicSliceOutput should be ok if the capacity is enough, and for each buffer we only do one range check.
+        // TODO: size should include channelCount etc
+        int sizeInBytes = getSizeInBytes();
+        byte[] finalSerializedBytes = new byte[sizeInBytes];
+        Slice slice = Slices.wrappedBuffer(finalSerializedBytes);
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java]: This can be simplified to `Slice slice = Slices.wrappedBuffer(new byte[getSizeInBytes()]);`, but why not use DynamicSliceOutput?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+public abstract class BlockEncodingBuffers
+{
+    protected int[] nextBatchOfPositions;
+    protected int nextBatchSize;
+    protected int bufferedRowCount;
+
+    abstract void prepareBuffers();
+
+    abstract void appendLongValues(long[] values);
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java]: BlockBuilders have methods for many types, e.g. writeByte, writeShort, writeInt. Are you expecting to add apeendBytes, appendShorts and appendInts?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+package com.facebook.presto.spi.block;
+
+import io.airlift.slice.SliceOutput;
+
+import static java.lang.Math.max;
+
+public abstract class BlockEncodingBuffers
+{
+    protected int[] nextBatchOfPositions;
+    protected int nextBatchSize;
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java]: I'd rename these to positions and positionCount.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    abstract int getSizeInBytes();
+
+    public void updateBufferedRowCount(int newRowCount)
+    {
+        bufferedRowCount += newRowCount;
+    }
+
+    public void copyPositions(int[] positions, int startIndex, int batchSize)
+    {
+        if (nextBatchOfPositions == null || nextBatchOfPositions.length < batchSize) {
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java]: extract this logic into a helper method to avoid copy-paste:

```
    private void ensurePositionsCapacity(int length)
    {
        if (nextBatchOfPositions == null || nextBatchOfPositions.length < length) {
            nextBatchOfPositions = new int[max(nextBatchOfPositions.length * 2, length)];
        }
    }
```

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockEncodingBuffers.java [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+    public static final int DEFAULT_MAX_ELEMENT_COUNT = 512 * 1024;
+
+    // TODO: These are single piece buffers for now. They will become linked list of pages requested from buffer pool.
+    // TODO: we need to decide how we can operate on the buffers. Slice itself doesn't allow appending, and DynamicSliceOutput does capacity check for every value. If we modify DynamicSliceOutput we need to modify Airlift.
+    // If we directly operate on byte[] we need to decide the interface. Currently we'll use ByteArrayUtils for updating batch of rows, and ByteArrays in Airlift to update single value
+    private byte[] nullsBuffer;
+    private int nullsBufferIndex;  //The next byte address if new values to be added.
+    private byte[] offsetsBuffer;
+    private int offsetsBufferIndex;  //The next byte address if new offset to be added.
+    private BlockEncodingBuffers rawBlockBuffer;
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockEncodingBuffers.java]: I can't find a place that initializes rawBlockBuffer. Would you give me a pointer?

**Comment by oerling** [comment]
It is not really practical to preserve RLE or dictionary across repartitioning.

The expectation is that for one message to the destination, you have tens or hundreds of input pages. The preserving stops applying whenever a dictionary is followed by a different dictionary.  You would have to make a combined dictionary which would be a lot of processing. Or you’d have to flush the previous content and then you’d get dictionaries of which only a few elements were referenced, which would negate the usefulness.

 

You can preserve these things better if you have a single destination but that is not where the volume is, so this kind of gain is likely below measurable.

 

So the whole deal is to just move data in bulk and loop over raw arrays.

 

For memory dynamics the best is probably to have the outer loop over columns and the inner loop over destinations. Scattered read and contiguous write.  The Aria sample works this way and is very fast. The Virtuoso equivalent is also like this.

 

The decisive factor is getting the data moved with the least instructions.

 

To save network, one could do a very quick and dirty attempt at making an RLE or dictionary from the per-destination buffer.  It is very fast to check for the possibility of dictionary once all the data is known. It is a lot harder to determine if two consecutive dictionaries are compatible.  The dictionary check could be for strings, something like if there are few distinct 8 byte prefixes then making a dictionary could be tried. Counting distinct prefixes goes with a Bloom filter and takes no time.  But I’m not aware of anybody actually doing this.

 

So in conclusion, I’d say to stick too minimizing the instruction count.

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java (hunk 5 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java]: I think reset() is too generic and reader might think this also resets the buffers. I renamed it to resetPositionCount() for now, and we can revisit the naming later.

**Comment by mbasmanova** [comment]
@yingsu00 @oerling Ying, Orri, one more question. Would it be possible to achieve reduction in instruction count by introducing batch APIs to BlockBuilder and using these in RepartitioningOperator?

```
/**
 * Writes the specified positions in the values array to the builder.
 */
BlockBuilder writeLongs(long[] values, int[] positions, int positionCount);
```

**Comment by oerling** [comment]
There would be some reduction that way but this would make the same amount of garbage, an amount equal to the data volume passing through.

So, to do a competent job of this one simply must write selected positions of the source to the output buffer. Everybody does this. This is the simplest of all operators. So doing a reasonable job of this is like the Most obvious of all the basics.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java (hunk 16 above)
Comment (yingsu00) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java]: DynamicSliceOutput.writeXXX() call Slices.ensureSize() every time. But for us we know the sizes in advance and don't need to do it. It's mainly a performance consideration

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java (hunk 18 above)
Comment (yingsu00) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java]: I renamed the positions. But I've seen OrcRecordReader uses nextBatchSize. I think nextBatchSize is more explanatory, and it avoids the confusions with the bufferedPositionCount

File: presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockEncodingBuffers.java (hunk 20 above)
Comment (yingsu00) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockEncodingBuffers.java]: Oops I removed it and forgot to add it back. Thanks for catching.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java (hunk 16 above)
Comment (mbasmanova) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java]: @yingsu00 Can this be solved by adding batch APIs to DynamicSliceOutput?

**Comment by mbasmanova** [comment]
@oerling 

> There would be some reduction that way but this would make the same amount of garbage, an amount equal to the data volume passing through.

Are you saying that efficiency gains are expected to come from reduced GC activity?

**Comment by oerling** [comment]
It is 1/3from less GC and 2/3 from fewer instructions or close enough.

You can look at the cpu and alloc profiles  in ariaprogress.rtf.

 

 

 

 

Basmanova <notifications@github.com> 
Sent: Friday, March 29, 2019 7:26 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] WIP - Repartitioning improvement (Do not review yet) (#12535)

 

@oerling <https://github.com/oerling> 

There would be some reduction that way but this would make the same amount of garbage, an amount equal to the data volume passing through.

Are you saying that efficiency gains are expected to come from reduced GC activity?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffersFactory.java (hunk 10 above)
Comment (yingsu00) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffersFactory.java]: @mbasmanova What about moving the BlockEncodingBuffersFactory as a static nested class in PartitionedOutputOperator, side by side with PartitionData? PartitionData has a PageSerde, and PageSerge as BlockEncodingSerde, then there are two classes implementing BlockEncodingSerde. Adding createBlockEncodingBuffers() there might complicate the code unnecessarily.

Comment (yingsu00) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffersFactory.java]: @mbasmanova Please excuse my previous response. I moved it to BlockEncodingSerde.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java (hunk 14 above)
Comment (yingsu00) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java]: @mbasmanova I don't quite understand. Could you elaborate why we need to remove this line?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java (hunk 17 above)
Comment (yingsu00) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java]: I changed it to appendFixedWidthValues() in the update.

**Comment by yingsu00** [comment]
> * I remember seeing repartitioning CPU dominating the overall CPU of the query when input is dictionary and/or run-length encoded. In these cases, repartitioning operator flattened the data greatly increasing its size. This change doesn't fix that, does it? How would you compare performance gains that could be realized from preserving the encoding vs. this change?

@mbasmanova Where did you see the case "repartitioning CPU dominating the overall CPU of the query when input is dictionary and/or run-length encoded" ? Do you have a query that can reproduces this? The existing Presto implementation encodes the DictionaryBlock by first compacting it, encoding the ids and dictionaries so I don't know quite understand how this happens. Aria-research implementation would flatten the DictionaryBlock though.

Currently I plan to keep previous Presto main's behavior, ie. encode/decode the Dictionary/RLE blocks in the same way. But if there are evident performance gains we will consider flattening them.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java (hunk 14 above)
Comment (mbasmanova) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java]: @yingsu00 capacity is always set to the same fixed value, hence, it is better to initialize it one and make capacity a final variable.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java [hunk 21]
```diff
@@ -295,4 +295,12 @@ default Block getLoadedBlock()
     {
         return this;
     }
+
+//    int[] getSizesInBytes();
+//
+//    int[] getSizesInBytes(int[] offsets, int offset, int length);
+
+    void appendPositionSizesInBytes(int[] sizesInBytes);
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java]: What's the semantic of these APIs ? :)

Comment (yingsu00) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java]: @wenleix These two methods populated the row sizes in the passed in int[] sizesInBytes for each row. We pass in this int[] to avoid memory allocation as much as possible.

**Comment by yingsu00** [comment]
Closing in favor of #13032
